### PR TITLE
KAFKA-13785: [7/N][Emit final] emit final for sliding window

### PR DESCRIPTION
**Pull request by lihaosky**
### Description
Implementation to emit final for sliding window agg. This PR is on top of https://github.com/apache/kafka/pull/11896

### Testing
pending unit test and integration test

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java [hunk 1]
```diff
@@ -48,15 +51,25 @@
...
     private boolean sendOldValues = false;
 
     public KStreamWindowAggregate(final Windows<W> windows,
                                   final String storeName,
                                   final Initializer<VAgg> initializer,
                                   final Aggregator<? super KIn, ? super VIn, VAgg> aggregator) {
+        this(windows, storeName, false, initializer, aggregator);
+    }
+
+    public KStreamWindowAggregate(final Windows<W> windows,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:74]: The PR title says "sliding window" but this it the "time window" class (ie, hopping/tumbling).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java [hunk 2]
```diff
@@ -50,17 +59,27 @@
...
     private final Aggregator<? super KIn, ? super VIn, VAgg> aggregator;
+    private final EmitStrategy emitStrategy;
 
     private boolean sendOldValues = false;
 
     public KStreamSlidingWindowAggregate(final SlidingWindows windows,
                                          final String storeName,
                                          final Initializer<VAgg> initializer,
                                          final Aggregator<? super KIn, ? super VIn, VAgg> aggregator) {
+        this(windows, storeName, EmitStrategy.onWindowUpdate(), initializer, aggregator);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java:57]: Would it be cleaner to remove this constructor and force the caller to alway pass in an `EmitStrategy`?

(Maybe same for hopping/tumbling windows?)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java [hunk 3]
```diff
@@ -81,23 +100,54 @@ private class KStreamSlidingWindowAggregateProcessor extends ContextualProcessor
...
+        private InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext;
+        private final TimeTracker timeTracker = new TimeTracker();
+        private final Time time = Time.SYSTEM;
 
         @Override
         public void init(final ProcessorContext<Windowed<KIn>, Change<VAgg>> context) {
             super.init(context);
-            final InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext =
-                (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
+            internalProcessorContext = (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java]: This `init()` code looks very alike to the hopping/tumbling window case -- might it be worth to add a new abstract class between `KStreamAggProcessorSupplier` and `KStreamSlidingWindowAggregate/KStreamWindowAggregate` that unifies this code?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java [hunk 4]
```diff
@@ -467,6 +516,72 @@ private boolean rightWindowIsNotEmpty(final ValueAndTimestamp<VAgg> rightWinAgg,
             return rightWinAgg != null && rightWinAgg.timestamp() > inputRecordTimestamp;
         }
 
+        private void maybeForwardUpdate(final Record<KIn, VIn> record,
+                                  final Window window,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java]: nit indention

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java [hunk 5]
```diff
@@ -467,6 +516,72 @@ private boolean rightWindowIsNotEmpty(final ValueAndTimestamp<VAgg> rightWinAgg,
...
+            if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+                return;
+            }
+            tupleForwarder.maybeForward(
+                record.withKey(new Windowed<>(record.key(), window))
+                    .withValue(new Change<>(newAgg, sendOldValues ? oldAgg : null))
+                    .withTimestamp(newTimestamp));
+        }
+
+        private void maybeForwardFinalResult(final Record<KIn, VIn> record, final long closeTime) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java]: as above: `maybeForwardFinalResult` (and actually also `maybeForwardUpdate` IIRC) look very similar to hopping/tumbling window code. Can we unify it?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SlidingWindowedKStreamImpl.java [hunk 6]
```diff
@@ -224,12 +227,20 @@ private <VR> StoreBuilder<TimestampedWindowStore<K, VR>> materialize(final Mater
...
-                        Duration.ofMillis(windows.timeDifferenceMs()),
-                        false
-                    );
+                    supplier = emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE ?
+                        RocksDbIndexedTimeOrderedWindowBytesStoreSupplier.create(
+                            materialized.storeName(),
+                            Duration.ofMillis(retentionPeriod),
+                            Duration.ofMillis(windows.timeDifferenceMs()),
+                            false,
+                           false
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SlidingWindowedKStreamImpl.java]: nit indention

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java (hunk 2 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java:57]: We can do that if there's no external caller.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java (hunk 3 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java]: I can take a look how to do it.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java (hunk 5 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java]: I'll take a look if it can be unified.

**Comment by guozhangwang** [comment]
https://github.com/apache/kafka/pull/12127

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java:57]: This function is now used by the co-group API which does not add emit strategy yet, so I think we should still keep it for now.

**Comment by guozhangwang** [comment]
Filed https://github.com/apache/kafka/pull/12135 (the actual diff after resolution is smaller).

**Pull request closed without merging**